### PR TITLE
Updates to evac merchant inventory and reloaded ammo itemgroups

### PR DESCRIPTION
**Pull request by chaosvolt**
#### Summary

SUMMARY: Balance "Allow blackpowder loads to spawn in reloaded ammo itemgroups, evac merchant inventory"

#### Purpose of change

One of the misc. ideas that was suggested I could tackle was to introduce more handloads, especially blackpowder ammo, to the evac merchant's inventory. I felt that I could improve this further by not just adding more variety to the evac merchant's ammo selection, but do so in a way that also allows blackpowder loads to potentially spawn in other locations too, along with making it easier to add them to other NPCs in the future.

Along the way this also turned into having to fix some problems where a lot of the evac merchant's stuff seemed to almost never spawn, most notably ammo. Even when I tested ammo spawns of 100% it was failing on me, so I had to look for an NPC shopkeep whose inventory DID seem to be working and update the evac merchant's inventory accordingly.

#### Describe the solution

1. Defined blackpowder ammo itemgroups for pistol, rifle, and shotgun loads. Same weights as their normal reloaded counterparts, minus any weights for entries lacking 
2. Added the blackpowder ammo groups to be cited by the reloaded itemgroups, with a current weight set to roughly 25% of the total weight in each group. Setting them aside as a separate itemgroup nested within the standard reloaded itemgroup makes things less messy and allows citing the blackpowder itemgroup elsewhere if desired in the future.
3. Reworked the inventory of the evac merchant. In a nutshell, during testing I discovered that the collection's item spawns were somehow being broken. In particular, even at 100% `prob` the ammo would almost never spawn. I had to resort to changing the evac merchant's inventory to work the same way the Hub 01 intercom merchant's stuff is set up, as I found that didn't seem to be having the same problem. The probabilities worked more or less decent as-is translated from percent chances to weights, but it could probably use some compact of the misc stuff into calls to gear/clothing itemgroups you'd find those items in.
4. MISC: Fixed .45-70 blackpowder load not having blackpowder and muzzle smoke flags, as I discovered it was still marked as "to do once that feature is merged" while checking on it due to noticing it had a non-standard ID compared to all other BP loads. Also changed its `proportional` block to be consistent with all other blackpowder loads.
5. MISC: Fixed `ammo_shotgun_reloaded` lacking .410 handload in it.

#### Describe alternatives you've considered

1. Just listing reloaded and blackpowder variants of the ammo originally cited in the itemgroup would've been the easiest way to do it, but I wouldn't have discovered how wonky the evac merchant's item spawns were behaving if I'd gone for that option first.
2. Since the suggestion that started this was "black powder loads in the evac merchant's inventory" I could also add separate calls to the BP itemgroups I added, to increase the relative weight of blackpowder ammo in the evac merchant's inventory.

#### Testing

1. Ported JSON changes to build 1834, load-tested.
2. Started in evac center multiple times, taking a look at contents of the merchant's inventory, both initially and after advancing time by a week.

#### Additional context

Some minor todo things that came up that can be done in a separate PR:
1. .45 Long Colt lacks a blackpowder variant, odd since it's one of the few cartridges in the game that historically originated before smokeless powder.
2. .410 shot also lacks a blackpowder variant.
3. 7.62x54mmR could use a blackpowder variant as well.
3. Will want to add weight entries to relevant blackpowder itemgroups once the items and recipes are added.

On top of that, further rebalances to the weights of the evac merchant's inventory might warrant ANOTHER follow-up PR.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
The new variant does a lot of repetition:

Screenshot from 2021-05-29 18-09-19

That merchant also has 5 wood axes, 3 pickaxes, 4 tactical vests, 3 army helmets and so on.
It's good for them to have some useful items, but guaranteed "kit" is a bit much, especially considering how easy it is to cheese merchants by selling them stupidly overpriced items for their modestly priced good stuff.
This is especially important in case of good quality ammo - they should have lower chances of getting non-reloaded and non-blackpowder rifle ammo.

Holy shit, reloaded ammo prices are broken: they use price on load, not price on finalization. That's big, gotta fix.

**Comment by chaosvolt** [comment]
> That merchant also has 5 wood axes, 3 pickaxes, 4 tactical vests, 3 army helmets and so on.

Sounds like the weight issue is higher-priority than I thought.

Any idea as the cause of the problem that led me to try this method? Shopkeepers using collections as-is seems to be buggy in ways I don't fully understand. Case in point, how the evac merchant was almost never stocking any ammo even when I bumped the prob up to 100.

I'd much rather be able to figure out what caused that and have it fixed so the merchant can still use the collection they originally used, because otherwise it's going to involve a lot of weight tweaks.

On the other hand...the root itemgroup the evac uses in this PR is STILL a collection, which eventually grabs ammo from a distribution. The old method also involves itemgroups in a collection that lead to picking ammo out of a distribution, so I don't quite get why one method works and the other doesn't. But it does mean hypothetically, I might be able to mess around with the current implementation to minimize how many gear items it calls for.

**Comment by Coolthulhu** [comment]
> Any idea as the cause of the problem that led me to try this method? Shopkeepers using collections as-is seems to be buggy in ways I don't fully understand. Case in point, how the evac merchant was almost never stocking any ammo even when I bumped the prob up to 100.

Holy shit, that's ugly: it seems that after 10th item, every next item has a chance of being the last one.
```
    int count = 0;
    bool last_item = false;
    while( shop_value > 0 && total_space > 0_ml && !last_item ) {
        item tmpit = item_group::item_from( from, calendar::turn );
        if( !tmpit.is_null() && total_space >= tmpit.volume() ) {
            tmpit.set_owner( *this );
            ret.push_back( tmpit );
            shop_value -= tmpit.price( true );
            total_space -= tmpit.volume();
            count += 1;
            last_item = count > 10 && one_in( 100 );
        }
    }
```

That's stupid.
If you want, you can try removing that weird limitation, the `last_item` and `count` variables and see what happen.

**Comment by chaosvolt** [comment]
Well that explains that. You want me to tackle that now, or should I move the ammo definitions up to the start of the collection and tackle this in a separate PR?

**Comment by Coolthulhu** [comment]
Can be here, though separate would be better.
Still, you should test both at the same time, so you could PR the other one and have it merged into this one. This way if this one is held up due to balance concerns, the `count` limitation removal could still be merged.

**Comment by chaosvolt** [comment]
I'll start up a separate PR for that, then rework this one as well, in a bit. Sorry for the delay, been out today.

**Comment by chaosvolt** [comment]
Welp, I'll have that change PR'd as soon as I stop getting access violations on the first game turn whenever I compile it in VS, and can thus actually properly playtest that change. The change itself compiles without any issue, and I'm getting the access violations even without the change, so it's something fucky with VS I suspect.

**Comment by chaosvolt** [comment]
Further update: I was able to get around that issue by teleporting instead of walking, and was able to confirm that changing this:
```
    int count = 0;
    bool last_item = false;
    while( shop_value > 0 && total_space > 0_ml && !last_item ) {
        item tmpit = item_group::item_from( from, calendar::turn );
        if( !tmpit.is_null() && total_space >= tmpit.volume() ) {
            tmpit.set_owner( *this );
            ret.push_back( tmpit );
            shop_value -= tmpit.price( true );
            total_space -= tmpit.volume();
            count += 1;
            last_item = count > 10 && one_in( 100 );
        }
    }
```
To this:
```
    while( shop_value > 0 && total_space > 0_ml ) {
        item tmpit = item_group::item_from( from, calendar::turn );
        if( !tmpit.is_null() && total_space >= tmpit.volume() ) {
            tmpit.set_owner( *this );
            ret.push_back( tmpit );
            shop_value -= tmpit.price( true );
            total_space -= tmpit.volume();
        }
    }
```

...had no apparent effect on this bug. Even after I injected ammo spawns with 100% prob, they wouldn't show up.

**Comment by chaosvolt** [comment]
Also noticed, the status quo for the evac merchant also does the "tons of redundant useful tools" thing due to a lot of the itemgroup calls having a count of 1-3 in them. In the meantime, I'm going to rework how the ammo spawns are placed to see if that ensures they're spawn before any tools spawn, overhaul their probabilities, then re-convert the evac merchant's inventory to be a standard collection, minus the `count` entries.

File: data/json/itemgroups/Weapons_Mods_Ammo/ammo.json [hunk 1]
```diff
@@ -295,15 +352,30 @@
...
       { "item": "reloaded_shot_slug", "prob": 30 },
-      { "item": "shot_scrap", "prob": 80 }
+      { "item": "shot_scrap", "prob": 80 },
+      { "item": "reloaded_410shot_000", "prob": 25 }
+    ]
+  },
+  {
+    "type": "item_group",
+    "id": "ammo_shotgun_blackpowder",
+    "//": "Blackpowder shotgun ammo balanced according to rarity.",
```
Comment (Coolthulhu) [inline data/json/itemgroups/Weapons_Mods_Ammo/ammo.json:371]: Those should not be rated by rarity of the original spawns but by usefulness and ease of crafting.
Black powder loads are mostly the post-apocalyptic ones and so should not have much bird shot, but slugs should be more common.
Same for reloaded ones, but to a lesser degree.

File: data/json/itemgroups/Weapons_Mods_Ammo/ammo.json [hunk 2]
```diff
@@ -295,15 +352,30 @@
   {
     "type": "item_group",
     "id": "ammo_shotgun_reloaded",
-    "//": "Reloaded shotgun ammo balanced according to rarity.",
+    "//": "Reloaded shotgun ammo balanced according to rarity.  Blackpowder loads should comprise roughly 25% of total weight.",
```
Comment (Coolthulhu) [inline data/json/itemgroups/Weapons_Mods_Ammo/ammo.json:355]: Better to make it explicit and enforced: make two groups, one for smokeless powder, one for black.

File: data/json/npcs/refugee_center/surface_staff/NPC_free_merchant_shopkeep.json [hunk 3]
```diff
@@ -52,30 +52,36 @@
...
-      { "group": "supplies_hardware", "prob": 20, "count": [ 1, 3 ] },
-      { "group": "supplies_fuel", "prob": 20, "count": [ 1, 3 ] },
-      { "group": "supplies_metal", "prob": 20, "count": [ 1, 3 ] },
-      { "group": "supplies_mechanics", "prob": 20, "count": [ 1, 3 ] },
+      { "group": "ammo_rifle_common", "prob": 10, "charges": [ 10, 30 ] },
+      { "group": "ammo_rifle_reloaded", "prob": 10, "charges": [ 10, 30 ] },
+      { "group": "ammo_rifle_blackpowder", "prob": 30, "charges": [ 10, 30 ] },
+      { "group": "ammo_pistol_common", "prob": 10, "charges": [ 10, 30 ] },
+      { "group": "ammo_pistol_reloaded", "prob": 10, "charges": [ 10, 30 ] },
+      { "group": "ammo_pistol_blackpowder", "prob": 30, "charges": [ 10, 30 ] },
```
Comment (Coolthulhu) [inline data/json/npcs/refugee_center/surface_staff/NPC_free_merchant_shopkeep.json:60]: I don't see the shotgun group here.
Fresh rifle ammo shouldn't be as common as fresh pistol ammo or reloaded pistol ammo.

File: data/json/itemgroups/Weapons_Mods_Ammo/ammo.json (hunk 1 above)
Comment (chaosvolt) [reply data/json/itemgroups/Weapons_Mods_Ammo/ammo.json:371]: I can go ahead and implement that then, sure. Making two separate groups also makes it easier to get the weights right, too.

File: data/json/itemgroups/Weapons_Mods_Ammo/ammo.json (hunk 2 above)
Comment (chaosvolt) [reply data/json/itemgroups/Weapons_Mods_Ammo/ammo.json:355]: That'd make it easier to get the weighting correct, so sure.

File: data/json/npcs/refugee_center/surface_staff/NPC_free_merchant_shopkeep.json (hunk 3 above)
Comment (chaosvolt) [reply data/json/npcs/refugee_center/surface_staff/NPC_free_merchant_shopkeep.json:60]: The evac merchant didn't have any shotshells in their original inventory, hence why I only added rifle and pistol ammo, to mimic what they were originally spawning with. And I can go ahead and tweak the rates.

**Comment by chaosvolt** [comment]
I'll be out and about today so I'll have to get to these when I get home.

File: data/json/npcs/refugee_center/surface_staff/NPC_free_merchant_shopkeep.json (hunk 3 above)
Comment (Coolthulhu) [reply data/json/npcs/refugee_center/surface_staff/NPC_free_merchant_shopkeep.json:60]: Well, it's better for them to have those. Shotgun ammo is common "in nature", so traders should have it as well. It's also not as useful as rifle ammo.

**Comment by Coolthulhu** [comment]
Some of the rolls produce too much ammo, but it's acceptable, I guess.
Can you report the problem with item groups having too few items?

**Comment by chaosvolt** [comment]
Ah, will open an issue for it in a bit then.

**Pull request merged**